--- aufs-6.5.orig/fs/aufs/Makefile	2023-09-15 10:48:39.000000000 +0200
+++ aufs-6.5.orig/fs/aufs/Makefile	2023-09-24 19:13:04.047014230 +0200
@@ -1,9 +1,14 @@
 # SPDX-License-Identifier: GPL-2.0

+CONFIG_AUFS_FS = m
+AUFS_DEF_CONFIG = -DCONFIG_AUFS_FS_MODULE -UCONFIG_AUFS
+include ${src}/magic.mk
+include ${src}/config.mk
+export CONFIG_AUFS_FS
+
+EXTRA_CFLAGS += ${AUFS_DEF_CONFIG}
 include ${src}/magic.mk
-ifeq (${CONFIG_AUFS_FS},m)
 include ${src}/conf.mk
-endif
 -include ${src}/priv_def.mk

 # cf. include/linux/kernel.h
@@ -11,7 +16,7 @@
 ccflags-y += -DDEBUG
 # sparse requires the full pathname
 ifdef M
-ccflags-y += -include ${M}/../../include/uapi/linux/aufs_type.h
+ccflags-y += -include ${M}/../include/uapi/linux/aufs_type.h
 else
 ccflags-y += -include ${srctree}/include/uapi/linux/aufs_type.h
 endif
